File: 2516-take-k-of-each-character-from-left-and-right/2516-take-k-of-each-character-from-left-and-right.java
class Solution {
    public int takeCharacters(String s, int k) {
        int n = s.length();
        HashMap<Character,Integer> m = new HashMap<>();
        
        for(int i=0;i<n;i++){
            m.put(s.charAt(i),m.getOrDefault(s.charAt(i),0)+1);
        }
        
        if (m.getOrDefault('a', 0) < k || m.getOrDefault('b', 0) < k || m.getOrDefault('c', 0) < k) {
            return -1;
        }
        
        int i = 0;int j = 0;
        int mxLen = 0;
        
        while(j<n){
            m.put(s.charAt(j),m.get(s.charAt(j))-1);
            if(m.get(s.charAt(j)) >= k){
                mxLen = Math.max(mxLen,j-i+1);
                j++;
            }
            else{
                while(m.get(s.charAt(j)) < k){
                    m.put(s.charAt(i),m.get(s.charAt(i))+1);
                    i++;
                }
                j++;
            }
        }
        return n - mxLen;
    }
}

File: 2516-take-k-of-each-character-from-left-and-right/2516-take-k-of-each-character-from-left-and-right.cpp
class Solution {
public:
    int takeCharacters(string s, int k) {
        unordered_map<char,int> m;
        int n = s.size();
        for(int i=0;i<n;i++){
            m[s[i]]++;
        }
        
        if(m['a'] < k || m['b'] < k || m['c'] < k)
            return -1;
    
        int i=0,j=0;
        int mxLen = 0;  // maximum window length that we will not delete
        while(j<n){
            m[s[j]]--;
            if(m[s[j]] >= k){
                mxLen = max(mxLen,j-i+1);
                j++;
            }
            else{
                while(m[s[j]] < k){
                    m[s[i]]++;
                    i++;
                }
                j++;
            }
        }
        return n-mxLen;
    }
};